### PR TITLE
Fix UnitarySynthesis with target that contains custom gates

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

Since Qiskit/qiskit-terra#9175 merged the UnitarySynthesis pass has been updated to consider multiple different potential target bases and select the best performing synthesis output found. When UnitarySynthesis is provided a Target object this involves querying the target for the error rates and duration of the instructions used and determining which bases are valid. As part of this lookup the Operator for the 2q gates in the target is needed to determine the characteristics of the 2q gates. However, the code for doing this wasn't considering that some operations in the target might not have sufficient information to build a unitary representation of the operation. This could come up with the use of custom opaque gates in the target that don't have a definition defined or a matrix defined. This commit fixes this oversight and handles the case where building an Operator for the gate fails.

### Details and comments

Related to Qiskit/qiskit-ibm-runtime#726 (the underlying issue is in the qiskit-ibm-runtime package for not mapping the "ecr" string to the "ECRGate" object, so this won't close that issue).

### CODE REVIEW

File: qiskit/transpiler/passes/synthesis/unitary_synthesis.py [hunk 1]
```diff
@@ -635,11 +636,17 @@ def _decomposer_2q_from_target(self, target, qubits, approximation_degree):
         decomposers = []
 
         def is_supercontrolled(gate):
-            kak = TwoQubitWeylDecomposition(Operator(gate).data)
+            try:
+                kak = TwoQubitWeylDecomposition(Operator(gate).data)
+            except QiskitError:
+                return False
```
Comment (jakelishman) [inline qiskit/transpiler/passes/synthesis/unitary_synthesis.py:642]: I think it's just the `Operator(gate)` call that should be permitted to fail here, right?  Can we just try/catch that instead, so we don't accidentally hide errors raised by `TwoQubitWeylDecomposition`?

Comment (mtreinish) [reply qiskit/transpiler/passes/synthesis/unitary_synthesis.py:642]: That's fair it is just the operator construction that we need to catch here. I'll update this accordingly.

Comment (mtreinish) [reply qiskit/transpiler/passes/synthesis/unitary_synthesis.py]: Updated in: https://github.com/Qiskit/qiskit-terra/pull/9677/commits/2a066bc12d33b9130bb1af381df78dc530508e37

**Review by jakelishman — APPROVED** [review]

**Comment by jakelishman** [comment]
@Mergifyio requeue

**Comment by jakelishman** [comment]
@Mergifyio refresh

**Pull request merged**